builtin/plugins/coin/coin.go

```diff
@@ -213,6 +257,95 @@ func mint(ctx contract.Context, to loom.Address, amount *loom.BigUInt) error {
 	return ctx.Set(economyKey, econ)
 }
 
+//Mint : to be called by CoinPolicyManager Responsible to mint coins as per various parameter defined
+func Mint(ctx contract.Context) error {
+	div := loom.NewBigUIntFromInt(10)
+	div.Exp(div, loom.NewBigUIntFromInt(18), nil)
+	var policy Policy
+	var amount *common.BigUInt
+	err := ctx.Get(policyKey, &policy)
+	if err != nil {
+		return errUtil.Wrap(err, "Failed to Get PolicyInfo")
+	}
+	changeRatioNumerator := loom.NewBigUIntFromInt(int64(policy.ChangeRatioNumerator))
+	changeRatioDenominator := loom.NewBigUIntFromInt(int64(policy.ChangeRatioDenominator))
+	blockHeight := loom.NewBigUIntFromInt(ctx.Block().Height)
+	totalSupply := loom.NewBigUIntFromInt(int64(policy.TotalSupply))
+	totalSupply.Mul(totalSupply, div)
+	blocksGeneratedPerYear := loom.NewBigUIntFromInt(int64(policy.BlocksGeneratedPerYear))
+	basePercentage := loom.NewBigUIntFromInt(int64(policy.BasePercentage))
+	operator := policy.Operator
+	_, modulus := big.NewInt(ctx.Block().Height).DivMod(big.NewInt(ctx.Block().Height), big.NewInt(int64(policy.BlocksGeneratedPerYear)), big.NewInt(int64(policy.BlocksGeneratedPerYear)))
+	year := blockHeight.Div(blockHeight, blocksGeneratedPerYear)
+	year = year.Add(year, loom.NewBigUIntFromInt(1))
+	var checkMintAmount = &types.BigUInt{
+		Value: *loom.NewBigUIntFromInt(0),
+	}
+	//Checks if minting amount is computed before or it is being computed for first time
+	err = ctx.Get(mintingAmountKey, checkMintAmount)
+	if err != nil && err != contract.ErrNotFound {
+		return err
+	}
+	//Minting Amount per block is set in ctx after minting amount is determined for year
+	if year.Cmp(loom.NewBigUIntFromInt(1)) == 0 {
+		//Determines minting amount at the beginning block of year or at block height at which minting is enabled
+		if modulus.Cmp(big.NewInt(1)) == 0 || err == contract.ErrNotFound {
+		//Minting Amount Computation for starting year
+			amount = totalSupply.Div(totalSupply, blocksGeneratedPerYear)
+			err = ctx.Set(mintingAmountKey, &types.BigUInt{
+				Value: *amount,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if modulus.Cmp(big.NewInt(1)) == 0 || err == contract.ErrNotFound {
+			//Operator Support to support different inflation ratio patterns for different year
+			if strings.EqualFold("div", operator) {
+				changeRatioDenominator = changeRatioDenominator.Mul(changeRatioDenominator, year)
+				basePercentage = basePercentage.Mul(basePercentage, changeRatioNumerator)
```

Comment:
```
   totalSupply = &econ.TotalSupply.Value
   inflationforYear := totalSupply.Mul(totalSupply, basePercentage)
   inflationforYear = inflationforYear.Div(inflationforYear, changeRatioDenominator)
   inflationforYear = inflationforYear.Div(inflationforYear, loom.NewBigUIntFromInt(100))
   amount = inflationforYear.Div(inflationforYear, blocksGeneratedPerYear)
   //Minting Amount per block computed starting from beginning block of that year and saved to avoid re-computations
   err = ctx.Set(mintingAmountKey, &types.BigUInt{
	Value: *amount,
   })
   if err != nil {
      return err
  }
  ```
  This is accomplished by this code, division is done in the end to avoid rounding off loss

Reply:
Change your terminology. TotalSupply means something else here.

Reply:
Should I rename Total Supply to BaseAmount as inflation is applied on top of BaseAmount.

Reply:
yeah

Reply:
Done.